packages/contracts-bedrock/src/Safe/LivenessGuard.sol

```diff
@@ -0,0 +1,135 @@
+
+contract LivenessGuard is SignatureDecoder, BaseGuard {
+    /// @notice Emitted when a new set of signers is recorded.
+    /// @param signers An arrary of signer addresses.
+    event SignersRecorded(bytes32 indexed txHash, address[] signers);
+
+    Safe public safe;
+    mapping(address => uint256) public lastSigned;
+
+    constructor(Safe _safe) {
+        safe = _safe;
+    }
+
+    /// @notice We just need to satisfy the BaseGuard interfae, but we don't actually need to use this method.
+    function checkAfterExecution(bytes32, bool) external pure {
+        return;
+    }
+
+    /// @notice Records the most recent time which any owner has signed a transaction.
+    function checkTransaction(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address payable refundReceiver,
+        bytes memory signatures,
+        address
+    )
+        external
+    {
+        require(msg.sender == address(safe), "LivenessGuard: only Safe can call this function");
+
+        // There are a number of ways in which we need to constrain this safe so that it cannot remove
+        // this guard, nor the LivenessModule.
+        // TODO(Maurelian): Figure out how to best address this. The following is just intended to outline the
+        // known mathods by which a Safe could remove the liveness checks.
+        // TODO(Maurelian): Do we _need_ to have this feature at all?
+        bytes4 dataSig = bytes4(data);
+        if (
+            to == address(safe)
+                && (dataSig == GuardManager.setGuard.selector || dataSig == ModuleManager.enableModule.selector)
+                || operation == Enum.Operation.DelegateCall
+        ) {
+            revert("LivenessGuard: cannot remove LivenessGuard or LivenessModule");
+        }
+
+        // This call will reenter to the Safe which is calling it. This is OK because it is only reading the
+        // nonce, and using the getTransactionHash() method.
+        bytes32 txHash = Safe(payable(msg.sender)).getTransactionHash(
+            // Transaction info
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            // Payment info
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver,
+            // Signature info
+            Safe(payable(msg.sender)).nonce() - 1
+        );
+        address[] memory signers = _getNSigners(txHash, signatures);
+        for (uint256 i = 0; i < signers.length; i++) {
+            lastSigned[signers[i]] = block.timestamp;
+        }
+        emit SignersRecorded(txHash, signers);
+    }
+
+    /// @notice Exctract the signers from a set of signatures.
+    function _getNSigners(bytes32 dataHash, bytes memory signatures) internal pure returns (address[] memory _owners) {
```

Comment:
Q from Matt: is there no way to get this directly from the Safe? ie. a public getter maybe?

Reply:
There does not appear to be any way.
By reading [the execution path](https://github.com/safe-global/safe-contracts/blob/e870f514ad34cd9654c72174d6d4a839e3c6639f/contracts/Safe.sol#L150-L176) in the Safe prior to the call to the guard, we can see nowhere that the addresses of the signers for the current transaction recorded in storage which would be necessary to make them accessible.